controllers/set/delete.go

```diff
@@ -45,30 +45,47 @@ func (c *controller) deleteFirewalls(r *controllers.Ctx[*v2.FirewallSet], fws ..
 
 	return nil
 }
-
-func (c *controller) deleteAfterTimeout(r *controllers.Ctx[*v2.FirewallSet], fws ...*v2.Firewall) ([]*v2.Firewall, error) {
+func (c *controller) deleteIfUnhealthyOrTimeout(r *controllers.Ctx[*v2.FirewallSet], fws ...*v2.Firewall) ([]*v2.Firewall, error) {
 	var result []*v2.Firewall
 
 	for _, fw := range fws {
-		fw := fw
+		if c.isFirewallUnhealthy(fw) {
```

Comment:
I think it would be clearer to split the handling depending on the phase because then the creation timeout and health timeout will not interfere. With some small changes on your newly introduced struct I think something like this would be clearer:

```golang
status := evaluateFirewallConditions(fw)

switch {
case status.CreateTimeout || status.HealthTimeout:
    r.Log.Info("firewall health or creation timeout exceeded, deleting from set", "firewall-name", fw.Name)

    err := c.deleteFirewalls(r, fw)
    if err != nil {
        return nil, err
    }

    result = append(result, fw)
}
```

The `isProgressing` that's used in `setStatus` would also not be required anymore as it can be derived in case all other cases do not match.

Reply:
I changed it, what do you think now?